### ISSUE TITLE
startWithAudioMuted and startWithVideoMuted not as expected from config

### ISSUE THREAD

**Issue by fipski** [jitsi/jitsi-meet#8144]
### Description:

Since commit [9c10ac3028548c3e8535d35789b23492b194e9fd](https://github.com/jitsi/jitsi-meet/commit/9c10ac3028548c3e8535d35789b23492b194e9fd) startWithAudioMuted and startWithVideoMuted could come from the local storage if the user had a previous setting. I would expect it to always come from config.startWithAudioMuted and config.startWithVideoMuted like it was before.

### Steps to reproduce:

1. set config.startWithAudioMuted to true in config.js
2. join a session
3. enable Audio
4. leave and join session again

### Expected behavior:

config.startWithAudioMuted and config.startWithVideoMuted control the mute setting when a conference is started

### Actual behavior:

Setting from local storage is taken into account.

### Server information:

Latest git on metal
Ubuntu 20.04

### Client information:

any

### Additional information:

I understand the reasons to take old user settings into account, but could this be done with a additional setting? Something like config.startWithAudioMutedFromStore? If enabled the default could be config.startWithAudioMuted and if the user has a setting in local sorage this could be taken into account? For me the current state creates a privacy issue.

**Comment by MDHollen** [comment]
Seconded - appending ?config.startWithAudioMuted=true&config.startWithVideoMuted=true has no impact on the default state of the video or audio when opening a call on Chrome, Firefox, or Edge. The commit mentioned by flipski renders these variables useless.

**Comment by saghul** [comment]
The config is loaded onto the store and extended with URL parameters, I think what might be happening here is that the config it's not being loaded early enough.

**Comment by paweldomas** [comment]
One solution to this problem would be to have a config option which disables storing of the muted states. Otherwise I think it makes sense for the local storage to override the config options.

**Comment by saghul** [comment]
IIRC the URL parameters will overrride the stored ones, but the config storage doesn't take place until a bit later I think.

**Comment by hristoterezov** [comment]
@MDHollen @saghul You need to append `#config.startWithAudioMuted=true&config.startWithVideoMuted=true` not `?config.startWithAudioMuted=true&config.startWithVideoMuted=true` and everything works as expected. We don't have an issue with the URL params.

About the original issue:
Note that the only thing that is not working is setting the values in the hosted config.js file itself. The URL params and the local storage values work as expected! 

Indeed the commit mentioned in the description results in changing the behaviour as described by @fipski but in the same time we need this for the prejoin screen to work properly.

@paweldomas I agree your solution will probably fix the problem in some cases but it will be a little bit limited because enabling the new config option will result in braking the prejoin screen.

Bottom line finding a solution to accommodate all these use cases is not that trivial! I don't know if we will have a solution soon.

**Comment by kiliankoller** [comment]
This is critical, as with the GDPR, it is required, that cionferences alway start with Audio and Video muten...

**Comment by amsnek** [comment]
can confirm, initial start on a clean browser, audio/video are "on" even if disabled via settings. It will "remember" the last setting on further visits but the config.js setting should be honored and/or overrule imo, especially for default.

**Comment by StrikerTwo** [comment]
Sorry, but what value does storing those settings locally have?
Now I always have to remember to mute my microphone and camera before leaving the conference, because otherwise they will be automatically active on the next one. This is not good user experience.

If you want to store something locally, add default settings and make them configurable on the start page. Then everybody can set the default for new conferences for themselves.

**Comment by saghul** [comment]
@amsnek 

> but the config.js setting should be honored and/or overrule imo, especially for default.

Yep, that's pretty much how we plan to fix it.

@StrikerTwo 

> Sorry, but what value does storing those settings locally have?
> Now I always have to remember to mute my microphone and camera before leaving the conference, because otherwise they will be automatically active on the next one. This is not good user experience.

Jitsi is intended to have video meetings, so allowing users to do that with the least amount of friction is the goal. Now, even if video is turned on when you join another meeting, you haven't actually joined it until you press join in what we call the "pre-join" screen, so your video is only visible to you.

> If you want to store something locally, add default settings and make them configurable on the start page. Then everybody can set the default for new conferences for themselves.

That's exactly what the pre-join page is for. It will start with your previous value and you can change it before actually joining.

Can you please elaborate on what problem you see with this approach, barring the bug that makes a default config of `true` not take precedence?

**Comment by StrikerTwo** [comment]
Simple: We don't use the pre-join page.

**Comment by saghul** [comment]
> Simple: We don't use the pre-join page.

I see. I assume you set `startWithVideoMuted` to `true` then in `config.js`, right? In that case the fix we intend to do will work for you as well, since a `true` value there will take precedence over any stored config.

**Comment by StrikerTwo** [comment]
That sounds good, thank you!

**Comment by cketti** [comment]
I'm using the following work-around until this bug is fixed: https://gist.github.com/cketti/f0ed9b722d04618b33a7269d030e1072

**Comment by setpill** [comment]
How does one enable the pre-join page?

**Comment by setpill** [comment]
Ah, found it, with `prejoinPageEnabled: true` in the `meet...config.js`.

**Comment by gmiller01** [comment]
Hi, I have upgrade our server from Ubuntu 18 to 20. First problem was the 'storage = none' what I fixed it with 'storage = memory'.
After that  I noticed that the camera and microphone are on by default. This is an unacceptable mistake. Can you imagine what happen, if 40 people connecting to a room with microphone switched on?  Babylonian chaos. 
Please fix that bug ASAP.

**Comment by saghul** [comment]
You can user startAudioMuted=2 and then every participant after the second will start muted.

**Comment by gmiller01** [comment]
> You can user startAudioMuted=2 and then every participant after the second will start muted.

Hi @saghul,
I have try it, but doesn't work in  /etc/jitsi/meet/my.own.domain-config.js.
Uncomment the "startAudioMuted: 2,"   but nothing happen.
Is there any place where I have to set "startAudioMuted=2"?

**Comment by agross** [comment]
@gmiller01 I believe @saghul meant that you could add the option to your meeting URLs in the query string. Which I find a bit impractical. I would prefer to see this issue being fixed.

**Comment by saghul** [comment]
> > You can user startAudioMuted=2 and then every participant after the second will start muted.
> 
> 
> 
> Hi @saghul,
> 
> I have try it, but doesn't work in  /etc/jitsi/meet/my.own.domain-config.js.
> 
> Uncomment the "startAudioMuted: 2,"   but nothing happen.
> 
> Is there any place where I have to set "startAudioMuted=2"?
> 
> 

Yes this goes in config.js. You mean that even after setting that all participants come in unmuted?

**Comment by 2old4it** [comment]
I hope this can be fixed soon. I is an annoying "feature". Thanks for the effort.

**Comment by venimus** [comment]
A temporary workaround is using `startAudioMuted = 0` and `startVideMuted = 0`, although you get an extra notification and it takes a second to take affect. So best is to fix it though

**Comment by gmiller01** [comment]
> A temporary workaround is using `startAudioMuted = 0` and `startVideMuted = 0` in /etc/jitsi/meet/my.own.domain-config.js, although you get an extra notification and it takes a second to affect. So best is to fix it though

Sorry for the bad news, but this workaround doesn't work.

**Comment by amsnek** [comment]
can confirm, that workaorund has no effect

**Comment by Semisol** [comment]
Still having problems with this. Any fixes yet?

**Comment by Semisol** [comment]
There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).

Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
```html
    <script>
      try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
    </script>
```

**Comment by gmiller01** [comment]
> There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).
> 
> Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
> 
> ```
>     <script>
>       try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
>     </script>
> ```

Thank you very much! You are my savior :)

**Comment by Semisol** [comment]
> > There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).
> > Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
> > ```
> >     <script>
> >       try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
> >     </script>
> > ```
> 
> Thank you very much! You are my savior :)

No problem :)

**Comment by migonr77** [comment]
Hi, is this (startWithVideoMuted: true) fixed in latest stable? It seems not to me. :( Thank you!

**Comment by saghul** [comment]
No, it's not fixed yet.

**Comment by Semisol** [comment]
I do not think it is. For now you can use this workaround:

> There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).
> 
> Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
> 
> ```
>     <script>
>       try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
>     </script>
> ```

**Comment by amsnek** [comment]
> I do not think it is. For now you can use this workaround:
> 
> > There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).
> > Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
> > ```
> >     <script>
> >       try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
> >     </script>
> > ```

hmm does this still work in latest stable release (2.0.5765)? for me it doesn't  work, havent used it before. Prejoin Page is disabled.

**Comment by Semisol** [comment]
> > I do not think it is. For now you can use this workaround:
> > > There is a workaround that I made that overwrites unsets the local config for video and audio muted (only if the prejoin page is disabled).
> > > Edit your `/usr/share/jitsi-meet/index.html`, and add the following code after the `<script><!--#include virtual="/config.js" --></script>` line:
> > > ```
> > >     <script>
> > >       try {let o = JSON.parse(localStorage.getItem("features/base/settings"));if (!o.prejoinPageEnabled){delete o.startWithAudioMuted; delete o.startWithVideoMuted};localStorage.setItem("features/base/settings", JSON.stringify(o))}catch{}
> > >     </script>
> > > ```
> 
> hmm does this still work in latest stable release (2.0.5765)? for me it doesn't work, havent used it before. Prejoin Page is disabled.

I am not sure.

**Comment by mhoeher** [comment]
For me, putting that code in the `index.html` also did not work. In addition, we're using a docker-compose based setup (from [here](https://github.com/jitsi/docker-jitsi-meet)), where modifying that file is not expected.

However, what works in our case is putting the following to the end of `./jitsi-meet-config/web/config.js`:

```js
try {
    let o = JSON.parse(localStorage.getItem("features/base/settings"));
    o.startWithVideoMuted = true;
    localStorage.setItem("features/base/settings", JSON.stringify(o))
} catch {
}
```

**Comment by amsnek** [comment]
this worked for me, thanks!

**Comment by jacereda** [comment]
Why not change the default values so that both audio and video start muted (unless explicitly told otherwise via config or params)?

**Comment by cketti** [comment]
I'd like to work on a fix. 

What is the intended behavior?
From the naming of the options I'd assume audio/video should be disabled if `startWithAudioMuted`/`startWithVideoMuted` is `true` in any of these: url parameters, local storage, or config. That would fix the issue I think most people here are having (prejoin page disabled and `startWithAudioMuted`/`startWithVideoMuted` set to `true` in the config).

Was the intention to ignore the config value on the prejoin page, i.e. enable video even when `config.startWithVideoMuted = true`?

Does it even make sense to have `startWithAudioMuted`/`startWithVideoMuted` in local storage? During my (admittedly limited) experiments the values never changed after creation.

**Comment by damencho** [comment]
> Does it even make sense to have `startWithAudioMuted`/`startWithVideoMuted` in local storage? During my (admittedly limited) experiments the values never changed after creation.

It changes in preJoin

**Comment by Semisol** [comment]
> For me, putting that code in the `index.html` also did not work. In addition, we're using a docker-compose based setup (from [here](https://github.com/jitsi/docker-jitsi-meet)), where modifying that file is not expected.
> 
> However, what works in our case is putting the following to the end of `./jitsi-meet-config/web/config.js`:
> 
> ```js
> try {
>     let o = JSON.parse(localStorage.getItem("features/base/settings"));
>     o.startWithVideoMuted = true;
>     localStorage.setItem("features/base/settings", JSON.stringify(o))
> } catch {
> }
> ```

Sort of same as my solution, but did not think about config.js as a possibility. Will update my comment.

**Comment by konanast** [comment]
Is there any update on this nothing seems to work as expected. [Semisol solution](https://github.com/jitsi/jitsi-meet/issues/8144#issuecomment-828159086) works but not on always (you have to refresh the page to mute).
I think this is very important, and makes jitsi unusable legally in EU due to GDPR.

**Comment by KodoGit** [comment]
Html fix works.

**Comment by amsnek** [comment]
html only works for people clicking on "start new meeting". It does not work for people joining a link to a conference directly.

**Comment by MardariG** [comment]
any status here? why not to make disabled video by default, and let people to activate the camera, I don't get it.

**Comment by StrikerTwo** [comment]
We "solved" it by activating the pre-join page again.

**Comment by ignlazaro** [comment]
Thank you all very much, the prejoinPageEnabled solution works.

**Comment by amsnek** [comment]
prejoinPage is only a workaround though, a proper solution (honor the config setting or disable video by default) is still required. Not every Instance wants or has a prejoin Page.

**Comment by Ademuz** [comment]
I also have this problem.

on-premise instance, latest stable jitsi-meet
Server side config.js section:
prejoinPageEnabled: true
startAudioMuted: 0

Each users have this parameter in their meeting link
#config.startWithAudioMuted=true

And still some random users join the meeting with the microphone turned on.
I can't reproduce this problem in a test environment with 5-10 participants.
But I face this issue every day at real meetings with 40 participants (all 40 video is on), 1-3 always join with mic is on.

**Comment by Corinari** [comment]
we also experience the bug and cant workaround it yet. 
we really need a working solution due to GDPR

**Comment by Semisol** [comment]
> prejoinPageEnabled: true

might be because they turned on the microphone in the prejoin page?

**Comment by Ademuz** [comment]
> > prejoinPageEnabled: true
> 
> might be because they turned on the microphone in the prejoin page?

As far as I know, they don't do anything like that.
In my case, they join the "stationary" room using a shortcut where the parameters are specified.
Due to the parameters in the shortcut, their microphone is always turned off on the prejoin page.
They get to the prejoin page and click join the meeting.

Every time after the meeting, we looked at the settings of such users and did not see anything unusual. Everyone has the preview page turned on, the microphone is turned off, they click join the meeting and do nothing else.

This problem is reproduced every day, on random users. After 15+ users in the room, 1-3 random users are sure to join with the microphone turned on.

All users use Windows 10 Pro and Google Chrome 91.xx

**Comment by Semisol** [comment]
I think then they turned it on in the prejoin page for probably another meeting, then it persisted. From what I remember the localStorage takes priority

**Comment by Ademuz** [comment]
> I think then they turned it on in the prejoin page for probably another meeting, then it persisted. From what I remember the localStorage takes priority

Yes, but the next day they use the shortcut again, which again sets them the correct parameters on the prejoin page.
And somehow they are not affected by startAudioMuted: 0 in config.js

And again, I can't reproduce this problem on my own using 5-10 test users. (my own chome tabs) Whatever I do, everything works as it should.

But as soon as there are 15+ real users with a video, the problem is reproduced.

**Comment by postwarbreakout** [comment]
Any news on this? We upgraded from an old version yesterday and this issue appeared as we do not use the prejoin page. I'm a bit shocked that this is over a year old.

**Comment by MardariG** [comment]
> 

just enable  pre-join page

**Comment by MardariG** [comment]
> Any news on this? We upgraded from an old version yesterday and this issue appeared as we do not use the prejoin page. I'm a bit shocked that this is over a year old.

just enable pre-join page

**Comment by amsnek** [comment]
in the newer versions this setting is: 
startAudioOnly:true,

**Comment by 0xnor0** [comment]
Using the workaround via config.js breaks the Android-Apps. I can only join a conference, if there is no overwrite in the config.js. Does anyone else have the problem?

**Comment by postwarbreakout** [comment]
> Using the workaround via config.js breaks the Android-Apps. I can only join a conference, if there is no overwrite in the config.js. Does anyone else have the problem?

Yes, I switched to this workaround instead: https://gist.github.com/cketti/f0ed9b722d04618b33a7269d030e1072

**Comment by tj0xin** [comment]
its still not working, when is the fix coming?

**Comment by saghul** [comment]
This bug doesn't have high priority because there are very few reasons why someone would want to disable the prejoin screen, specially given the original reason why this was raised was privacy.

**Comment by vladertel** [comment]
Hi, @saghul! We (Code With Me at JetBrains) would really like having this one fixed. I'm tired of patching this every release

**Comment by Ashik55** [comment]
any solution for this issue??
startWithVideoMuted config is not working, its always opens the camera. I have disabled  pre-join page.

**Comment by saghul** [comment]
No update. We recommend you enable the pre-join page.

**Comment by Ashik55** [comment]
@saghul we are building a calling app (ios & android) with audio & video 2 buttons. is there any workaround to disable camera for **setAudioOnly** through any config or modification from jitsi side as we can't enable the **prejoinpage** due to UX issue.


```
        let userInfo = JitsiMeetUserInfo()
        userInfo.displayName = displayName // Set the desired display name here

        
        let defaultOptions = JitsiMeetConferenceOptions.fromBuilder { (builder) in
            builder.serverURL = URL(string: meetUrl)
            builder.userInfo = userInfo // Set the user info here
            
            //isAudioCall is a boolean for confirming if its an audio call or video call
            builder.setAudioOnly(self.isAudioCall)
            builder.setVideoMuted(self.isAudioCall)
            
            builder.setFeatureFlag("welcomepage.enabled", withValue: false)
            builder.setFeatureFlag("prejoinpage.enabled", withValue: false)
            
             // Try setting the audio device flag
             builder.setFeatureFlag("toolbox.alwaysVisible", withValue: false)
             builder.setFeatureFlag("reactions.enabled", withValue: false)
             builder.setFeatureFlag("chat.enabled", withValue: false)
            
        }
```

**Comment by Ashik55** [comment]
@saghul From the receiver’s end, it’s impossible to detect whether the call is AudioOnly or not. Additionally, when joining the call, the URL gets modified from
`https://meet.jitsi.com/ashik5575#config.startAudioOnly=true` to `https://meet.jitsi.com/ashik5575`, but the call still initiates as audio-only. 
However, the receiver always detects it as a regular video call, automatically opening the camera and loudspeaker, as there’s no way to detect the AudioOnly state after the URL modification.

How can I detect from the receiver’s end whether the call is AudioOnly or not?

**Comment by saghul** [comment]
The call is not audio only. That is how a given participant decides to join it. It's not enforced in any way.

**Comment by corby-github** [comment]
@saghul We are seeing this bug as well -  On Android using  when `.setFeatureFlag("prejoinpage.enabled", false)` - on initial install (uninstall, reinstall, join room) you will see the error.

You will be prompted for mic access (on join) and after granted, pressing the mic button will not work. Pressing the camera button will not prompt you for access.

On mic press, you will see:
```
ExternalAPI Sending event: CUSTOM_BUTTON_PRESSED with data: {"id":"microphone"}
```
but not:
```
ExternalAPI Sending event: AUDIO_MUTED_CHANGED with data: {"muted":true}
```


You will be prompted for voice even if its not enabled

I think the last time this worked, is around sdk version `9.2.2` - (I know, this is old, before the SDKs were unified) - 

I tested this today (4/15/25) with the latest `JitsiSDKTest` for Android (Java) against `https://meet.jit.si/...`.

To reproduce:
```
1. Grab the latest JitsiSDKTest code
2. Uninstall `JitsiSDKTest` (if installed)
3. Add `.setFeatureFlag("prejoinpage.enabled", false)` to the `defaultOptions`
4. Start a meeting in a browser
5. Build and launch the `JitsiSDKTest` app onto an Android device
6. Notice the permissions prompt for `mic` access on join
7. Grant access (while running app or 'just this time')
8. Tap on the mic button <--- you will see the error here
9. Also try the camera button (error here, too)
10. Leave meeting
11. Rejoin
12. Notice the permissions prompt for `notification` access on join (Grant access)
13. Try the mic or camera icon <--- you will see the error here, still
```

We did not notice this before because I had been testing upgrades, but with fresh installs, it happens all the time

**Comment by corby-github** [comment]
FWIW, As a work around to this, we started checking/prompting for permission before calling `JitsiMeetActivity.launch()` - This fixes Android first run issues

**Comment by Irfan762** [comment]
@fipski can i contibute in this project

**Comment by farapholch** [comment]
We are experiencing this behaviour when using Rocket.Chat Jitsi integration aswell

**Comment by damencho** [comment]
> We are experiencing this behaviour when using Rocket.Chat Jitsi integration aswell

What do you experience? Are you disabling prejoin and using that as URL parameter?

**Comment by farapholch** [comment]
> > We are experiencing this behaviour when using Rocket.Chat Jitsi integration aswell
> 
> What do you experience? Are you disabling prejoin and using that as URL parameter?

yes prejoin disabled

`- name: ENABLE_PREJOIN_PAGE value: "false"` and
`config.prejoinPageEnabled=false`

**Comment by YYSLansus** [comment]
reproducible here. I was using the docker jitis meet with version JITSI_IMAGE_VERSION=stable-11146-1. These five variables were configured in the .env file. However, when the user entered the meeting with url on the Google Chrome/Microsoft Edge, the microphone always was unmuted. 

START_WITH_AUDIO_MUTED=true
START_WITH_VIDEO_MUTED=true
START_AUDIO_MUTED=1
START_VIDEO_MUTED=1
ENABLE_PREJOIN_PAGE=0

**Comment by Yash1927** [comment]
Hi @saghul, @damencho,  I confirmed this still happens on current master. A saved `false` overrides `startWithAudioMuted: true` or `startWithVideoMuted: true` from `config.js`.

Can I open a PR that makes `config.js: true` override the saved setting, without changing URL overrides or prejoin settings?
